chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java

```diff
@@ -279,383 +313,398 @@ public static double getDouble(final IoBuffer readBuffer) {
+            }
+            if (buffer.position() < dataStartPosition) {
+                lastFieldHeader.setFieldDescription(buffer.getString());
+            }
+            if (buffer.position() < dataStartPosition) {
+                lastFieldHeader.setFieldDirection(buffer.getString());
+            }
+            if (buffer.position() < dataStartPosition) {
+                final String[] fieldGroups = buffer.getStringArray();
+                lastFieldHeader.setFieldGroups(fieldGroups == null ? Collections.emptyList() : Arrays.asList(fieldGroups));
+            }
+        } else {
+            buffer.position(dataStartPosition);
         }
-        return retVal;
-    }
 
-    public static int[] getIntArray(final IoBuffer readBuffer) {
-        if (readBuffer.hasRemaining()) {
-            return readBuffer.getIntArray();
+        // check for header-dataStart offset consistency
+        if (buffer.position() != dataStartPosition) {
+            final int diff = dataStartPosition - buffer.position();
+            throw new IllegalStateException("could not parse FieldHeader: fieldName='" + dataType + ":" + fieldName + "' dataOffset = " + dataStartOffset + " bytes (read) -- " //
+                                            + " buffer position is " + buffer.position() + " vs. calculated " + dataStartPosition + " diff = " + diff);
         }
-        throw new IndexOutOfBoundsException(READ_POSITION_AT_BUFFER_END);
-    }
 
-    public static int getInteger(final IoBuffer readBuffer) {
-        if (readBuffer.hasRemaining()) {
-            return readBuffer.getInt();
+        if (dataSize >= 0) {
+            return lastFieldHeader;
         }
-        throw new IndexOutOfBoundsException(READ_POSITION_AT_BUFFER_END);
-    }
 
-    public static <E> List<E> getList(final IoBuffer readBuffer, final List<E> collection) {
-        final DataType valueDataType = DataType.fromByte(readBuffer.getByte());
-
-        // read value vector
-        final Object[] values = getGenericArrayAsBoxedPrimitive(readBuffer, valueDataType);
-        final int nElements = values.length;
-        final List<E> retCollection = collection == null ? new ArrayList<>(nElements) : collection;
-        for (int i = 0; i < nElements; i++) {
-            retCollection.add((E) values[i]);
-        }
+        // last-minute check in case dataSize hasn't been set correctly
+        if (dataType.isScalar()) {
+            dataSize = dataType.getPrimitiveSize();
+        } else if (dataType == DataType.STRING) {
+            final int pos = buffer.position();
+            dataSize = FastByteBuffer.SIZE_OF_INT + buffer.getInt(); // <(>string size -1> + <string byte data>
+            buffer.position(pos);
+        }
+        lastFieldHeader.setDataSize(dataSize);
+
+        return lastFieldHeader;
+    }
+
+    @Override
+    public <E> List<E> getList(final List<E> collection, BiFunction<Type, Type[], FieldSerialiser<E>> serialiserLookup) {
```

Comment:
Codacy found an issue: [The method 'getList(List, BiFunction)' has an NPath complexity of 204, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5905529)